### PR TITLE
DAOS-14550 control: Prevent SPDK calls if disable_hugepages true in cfg

### PR DESCRIPTION
**Pull request by tanabarr**
Avoid calling into SPDK if disable_hugepages is set in the server
config file. This change prevents access during StorageScan,
StorageFormat, StorageNvmeRebind and StorageNvmeAddDevice server
ControlService handlers and when cleaning hugepages during start-up.

Also:
- Improve fault information messages
- Improve skip related messaging during format
- Cover missing config in storage command handler unit tests
- Fix quirks in storage command display output
- Improve unit test coverage of storage pretty printers

Features: control
Required-githooks: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by knard-intel — APPROVED** [review]
LGTM

**Review by mjmac — DISMISSED** [review]

**Comment by tanabarr** [comment]
If we want coverage for disable_hugepages in CI we probably want to run a SCM-only test after clearing hugepages on a host, that would have caught the `permission denied` SPDK errors. `sudo su -c "echo 0 > /proc/sys/vm/nr_hugepages"`

File: src/control/server/ctl_storage_rpc.go
Comment (knard-intel) [inline src/control/server/ctl_storage_rpc.go:36]:
```suggestion
	msgNvmeFormatSkipNotDone = msgNvmeFormatSkip + ", SCM was already formatted"
```

**Review by knard-intel — APPROVED** [review]
Mostly OK, just one small remarks regarding a confusing information message.

File: src/control/server/ctl_storage_rpc.go
Comment (tanabarr) [reply src/control/server/ctl_storage_rpc.go:36]: that Skip message is accurate, it represents the case when no formatting was performed, it could be because SCM was already formatted or because of something else

**Comment by tanabarr** [comment]
GATEKEEPER: Please use the PR title and description above as the commit message when landing. TIA

File: src/control/server/ctl_storage_rpc.go
Comment (knard-intel) [reply src/control/server/ctl_storage_rpc.go:36]: OK, I was understanding that it was only possible if the SCM was previously formatted.
Thanks for the clarification.

**Review by mjmac — COMMENTED** [review]
A few things:
  * First, a general grumble... If you're going to sneak in unrelated formatting changes that don't affect functionality, please put them in their own commit so that reviewers don't have to wade through pages of whitespace updates before getting to the "meat" of the patch.
  * Second, the general approach here seems to be to turn a number of function calls into no-ops if hugepages are disabled in the server config. Wouldn't it be better to just not call those functions? It would certainly be easier to understand.
  * Finally, the fact that the code needs to be updated in so many places indicates some problems with delegation of responsibility. Is it possible to restructure the code so that the behavior is controlled in one place? My concern is that having responsibility for this sprinkled around is harder to understand and maintain, and increases the odds that future changes will miss a spot.

**Comment by tanabarr** [comment]
> A few things:
> 
> * First, a general grumble... If you're going to sneak in unrelated formatting changes that don't affect functionality, please put them in their own commit so that reviewers don't have to wade through pages of whitespace updates before getting to the "meat" of the patch.
Sure
> * Second, the general approach here seems to be to turn a number of function calls into no-ops if hugepages are disabled in the server config. Wouldn't it be better to just not call those functions? It would certainly be easier to understand.
I will try to do that
> * Finally, the fact that the code needs to be updated in so many places indicates some problems with delegation of responsibility. Is it possible to restructure the code so that the behavior is controlled in one place? My concern is that having responsibility for t his sprinkled around is harder to understand and maintain, and increases the odds that future changes will miss a spot.
That sounds sensible and I have thought the same, as this is a reactive fix are you okay if I do that in a separate PR? I have to balance against other priorities for 2.6.

File: src/control/server/ctl_storage_rpc.go
Comment (kjacque) [inline src/control/server/ctl_storage_rpc.go:994]: nit - The reasoning for the skip is included in the result returned from formatNvme, right? Just wondering if we need this message at the Notice level if the user will be notified anyway.

File: src/control/server/faults.go
Comment (kjacque) [inline src/control/server/faults.go:63]: Just a suggestion for tightening up the naming/messaging:
```suggestion
	FaultHugepagesDisabled = serverFault(
		code.ServerHugepagesDisabled,
		"the use of hugepages has been disabled in the server config",
		"set false (or remove) disable_hugepages parameter in config and reformat storage, then retry the operation",
	)
```

File: src/control/server/instance_storage_rpc_test.go
Comment (kjacque) [inline src/control/server/instance_storage_rpc_test.go:188]: I wonder about changing this behavior... Does this have to be fatal? It is just a scan, right? Does the engine have to be joined for that?

**Review by kjacque — COMMENTED** [review]
If the change in behavior has already been discussed and thought through elsewhere, I would be willing to approve as-is. The other two items are minor and could be addressed in another PR.

File: src/control/server/instance_storage_rpc_test.go
Comment (kjacque) [reply src/control/server/instance_storage_rpc_test.go:188]: Discussed in a chat -- the "old" behavior here is a relatively new change, and logically we should not get into a position where the engine has a bad rank but allows a storage scan. This makes sense to me.

**Review by kjacque — APPROVED** [review]
Works for me. My suggestions, and greater improvements suggested by Mike, would be fine to address in a future PR.

**Comment by mjmac** [comment]
> That sounds sensible and I have thought the same, as this is a reactive fix are you okay if I do that in a separate PR? I have to balance against other priorities for 2.6.

That's fine with me. Please create a ticket so that the work isn't forgotten, though.

**Comment by tanabarr** [comment]
@mjmac does the most recent commit address your second point sufficiently? I'm assuming you don't want me to address the first point in this PR as it has already been reviewed and I will address your third point in another PR (and create a ticket for it). If it looks okay then I will update the unit tests appropriately.

**Comment by tanabarr** [comment]
> > That sounds sensible and I have thought the same, as this is a reactive fix are you okay if I do that in a separate PR? I have to balance against other priorities for 2.6.
> 
> That's fine with me. Please create a ticket so that the work isn't forgotten, though.

https://daosio.atlassian.net/browse/DAOS-15086

**Review by mjmac — COMMENTED** [review]
This looks better to me. Conceptually, it's much easier to understand smaller functions that either succeed at doing the thing or fail to do the thing rather than maybe doing the thing.

Holding off on approval for now, as there have been a number of changes in related areas and it would be prudent to re-test after landing to ensure that logic changes don't result in test failures.

File: src/control/server/faults.go
Comment (tanabarr) [reply src/control/server/faults.go]: done

**Comment by tanabarr** [comment]
GATEKEEPER: please use the PR title and description as commit message when landing, TIA

**Review by knard-intel — APPROVED** [review]

**Review by mjmac — APPROVED** [review]
LGTM

**Pull request merged**
